src/com/google/javascript/jscomp/ProductionCoverageInstrumentationCallback.java

```diff
@@ -137,56 +137,123 @@ private Node newInstrumentationNode(NodeTraversal traversal, Node node, String f
-      uniqueIdentifier = new ArrayList<>();
-      paramValue = new ArrayList<>();
+
+      paramValueEncodings = new HashMap<>();
+
+      // A LinkedHashMap is used so that when keys are printed, keySet() will obtain them in the
+      // insertion order which corroborates to the index. This helps to avoid the need of sorting
+      // by the Integer value and is more convenient.
+      fileNamesMap = new LinkedHashMap<>();
+      fileNamesCounter = 0;
+      functionNamesMap = new LinkedHashMap<>();
+      functionNamesCounter = 0;
+      typeMap = new LinkedHashMap<>();
+      typeCounter = 0;
     }
 
-    public String getEncodedParam(String fileName, String functionName, String type) {
-      String combinedParam = lenientFormat("%s %s %s", fileName, functionName, type);
-      long uniqueIdentifier = getUniqueIdentifier();
+    private String getEncodedParam(String fileName, String functionName, String type) {
 
-      if (uniqueIdentifier > Integer.MAX_VALUE) {
-        throw new ArithmeticException(
-            "Unique Identifier exceeds value of Integer.MAX_VALUE, could not encode with Base 64"
-                + " VLQ");
+      if (!fileNamesMap.containsKey(fileName)) {
+        fileNamesMap.put(fileName, fileNamesCounter);
+        fileNamesCounter++;
+      }
+      if (!functionNamesMap.containsKey(functionName)) {
+        functionNamesMap.put(functionName, functionNamesCounter);
+        functionNamesCounter++;
       }
+      if (!typeMap.containsKey(type)) {
+        typeMap.put(type, typeCounter);
+        typeCounter++;
+      }
+
+      Integer fnNameIndex = functionNamesMap.get(functionName);
+      Integer fileNameIndex = fileNamesMap.get(fileName);
+      Integer typeIndex = typeMap.get(type);
 
       StringBuilder sb = new StringBuilder();
 
       try {
-        Base64VLQ.encode(sb, Math.toIntExact(uniqueIdentifier));
+        Base64VLQ.encode(sb, fileNameIndex);
+        Base64VLQ.encode(sb, fnNameIndex);
+        Base64VLQ.encode(sb, typeIndex);
       } catch (IOException e) {
         throw new AssertionError(e);
       }
 
-      String result = sb.toString();
+      String encodedParam = sb.toString();
+
+      if (!paramValueEncodings.containsKey(encodedParam)) {
+        long uniqueIdentifier = getUniqueIdentifier();
+        if (uniqueIdentifier > Integer.MAX_VALUE) {
+          throw new ArithmeticException(
+              "Unique Identifier exceeds value of Integer.MAX_VALUE, could not encode with Base 64"
+                  + " VLQ");
+        }
 
-      addParamMapping(result, combinedParam);
+        sb = new StringBuilder();
+
+        try {
+          Base64VLQ.encode(sb, Math.toIntExact(uniqueIdentifier));
+        } catch (IOException e) {
+          throw new AssertionError(e);
+        }
+
+        paramValueEncodings.put(encodedParam, sb.toString());
+      }
 
-      return result;
+      return paramValueEncodings.get(encodedParam);
     }
 
     private long getUniqueIdentifier() {
```

Comment:
Let's rename this to `generateUniqueIdentifier`. `get` operations don't change state. Especially in java where getters are expected to be dumb accessors.